### ISSUE TITLE
[PROPOSAL] Improve testing

### ISSUE THREAD

**Issue by BSFishy** [opensearch-project/oui#731]
Currently, OUI has 2 types of testing: unit testing and accessibility testing. However, as a component library, it is important that we take into account at least 2 additional types of tests: visual regression testing and interaction testing.

### Visual Regression Testing

Visual regression testing is a way to track visual changes of components over time. This ensures that, for example, the layout of a component isn't broken when we make changes. It is difficult to see visual changes in a PR, especially when all we have is code to look at. This should give us screenshots of a before and after, as well as potentially a visual diff between the two.

In a perfect world, we could have an instance of the docsite spun up for each PR, and this is something we may look into in the future, however that's out of the scope of this issue. Having an instance of the docsite running would allow us to see and play with the docsite with those changes in a production-style environment. We could, additionally manually spin up an instance per PR, however that is a lot of time and effort on the reviewer's side, and a lot of wasted time if we can automate the system.

### Interaction Testing

Interaction testing is the testing of interactions with a component. For example, clicking, hovering, typing, etc. to ensure each component acts as it is meant to. This ensures that interactions aren't breaking between changes in a way that may be difficult to tell just by looking at code.

Like mentioned in the Visual Regression Testing section, this could be done by spinning up an instance of the docsite, however it is a lot more convenient to set up automations to complete this work for us.

## Libraries

There are a number of libraries and frameworks that can help us get to where we want to be. The main task of this issue will be comparing and contrasting the different options and figuring out which one we will want to use.

I have identified a few options that may be useful, however this is by no means an extensive list. Feel free to add any number to the list:

 1. Playwright
 2. Cypress
 3. Chromatic
 4. Puppeteer

We should be evaluating on a few criteria. This is by no means an exhaustive list. Feel free to add any more criteria you deem to fit:

 1. How active is the repository? When was the last release? When was the last commit? Are there open issues and PR's? Is there any activity on the issues and PR's?
 2. Does it provide visual regression testing?
 3. How is visual regression done? Is a React component passed to a function? Does a server need to be started and the tester pointed to a URL?
 4. How are visual regressions tracked? Is it an image file hosted in the repository? Is it hosted on a cloud service?
 5. How are visual regressions shown to the user? Is there a command output that tells the user? Is a diff image created?
 6. Does it provide interaction testing?
 7. How is interaction testing done? Is it a system like Jest where we define units and test specific things in the units?

## Follow up work

Once the initial research is done, we will need to implement the groundwork for the system. Most likely, this will involve setting up the configuration, creating a script to run the tests, and creating some sort of template for the tests. Once we have the groundwork, creating tests will be an excellent task for CCI contributors.

**Comment by anirudha** [comment]
Closing — this issue has been open for 2+ years without meaningful progress. The codebase has evolved significantly. If still relevant, please open a new issue with updated requirements and context.
